File: src/android/app/src/main/java/com/dogwalker/app/domain/usecase/walk/TrackWalkUseCase.kt
/**
 * Human Tasks:
 * 1. Verify coroutine scope configuration in the dependency injection setup
 * 2. Ensure proper error handling is implemented in the UI layer for location tracking failures
 * 3. Configure location permission handling in the UI layer before calling these methods
 * 4. Verify GPS accuracy settings match the application requirements
 */

package com.dogwalker.app.domain.usecase.walk

// Internal imports
import com.dogwalker.app.domain.model.Walk
import com.dogwalker.app.domain.model.Location
import com.dogwalker.app.data.repository.WalkRepository
import com.dogwalker.app.data.repository.LocationRepository

// Coroutines - version 1.6.4
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Use case class that provides business logic for tracking a walk in real-time.
 * 
 * Requirements addressed:
 * - Service Execution (1.3 Scope/Core Features/Service Execution)
 *   Implements live GPS tracking and route recording functionality for walks,
 *   providing methods to save location updates and retrieve walk routes.
 */
class TrackWalkUseCase(
    private val walkRepository: WalkRepository,
    private val locationRepository: LocationRepository
) {
    /**
     * Tracks the current location of a walk by saving the location data to the database.
     * This method should be called periodically during an active walk to record the route.
     *
     * Requirements addressed:
     * - Service Execution (1.3)
     *   Implements real-time GPS tracking by saving location updates during walks
     *
     * @param walkId Unique identifier of the walk being tracked
     * @param latitude Current latitude coordinate
     * @param longitude Current longitude coordinate
     * @param timestamp Unix timestamp when the location was recorded
     */
    suspend fun trackLocation(
        walkId: String,
        latitude: Double,
        longitude: Double,
        timestamp: Long
    ) {
        withContext(Dispatchers.IO) {
            try {
                // Validate input parameters
                require(walkId.isNotBlank()) { "Walk ID cannot be blank" }
                require(latitude in -90.0..90.0) { "Invalid latitude value" }
                require(longitude in -180.0..180.0) { "Invalid longitude value" }
                require(timestamp > 0) { "Invalid timestamp" }

                // Retrieve the walk entity
                val walk = walkRepository.getWalkById(walkId)
                requireNotNull(walk) { "Walk not found: $walkId" }

                // Create a new location object
                val location = Location(
                    id = "${walkId}_${timestamp}",
                    latitude = latitude,
                    longitude = longitude,
                    userId = walk.userId,
                    walkId = walkId,
                    timestamp = timestamp
                )

                // Save the location to the database
                locationRepository.saveLocation(location)

                // Update the walk's location list
                val updatedLocations = walk.locations + location
                val updatedWalk = walk.copy(locations = updatedLocations)
                walkRepository.insertWalk(updatedWalk)
            } catch (e: Exception) {
                // Log the error and rethrow
                println("Error tracking location for walk $walkId: ${e.message}")
                throw e
            }
        }
    }

    /**
     * Retrieves the complete route of a walk by fetching all associated location data.
     * This method can be used to display the walk route on a map or calculate statistics.
     *
     * Requirements addressed:
     * - Service Execution (1.3)
     *   Supports route recording by providing access to the complete walk route
     *
     * @param walkId Unique identifier of the walk
     * @return List of Location objects representing the walk's route
     */
    suspend fun getRoute(walkId: String): List<Location> {
        return withContext(Dispatchers.IO) {
            try {
                // Validate walk ID
                require(walkId.isNotBlank()) { "Walk ID cannot be blank" }

                // Verify that the walk exists
                val walk = walkRepository.getWalkById(walkId)
                requireNotNull(walk) { "Walk not found: $walkId" }

                // Retrieve all locations for the walk
                locationRepository.getLocationsForWalk(walkId)
            } catch (e: Exception) {
                // Log the error and rethrow
                println("Error retrieving route for walk $walkId: ${e.message}")
                throw e
            }
        }
    }

    companion object {
        /**
         * Minimum time interval (in milliseconds) between location updates
         */
        const val MIN_UPDATE_INTERVAL_MS = 5000L // 5 seconds

        /**
         * Maximum time interval (in milliseconds) between location updates
         */
        const val MAX_UPDATE_INTERVAL_MS = 30000L // 30 seconds
    }
}

File: src/ios/DogWalker/Domain/UseCases/Walk/TrackWalkUseCase.swift
// CoreLocation framework - Latest version
import CoreLocation

// MARK: - Human Tasks
/*
Prerequisites:
1. Configure location services in Info.plist with proper usage descriptions
2. Set up proper error handling for location tracking failures
3. Review and implement proper data retention policies for location data compliance
4. Configure network security settings for location data transmission
*/

// Internal imports using relative paths
import "../../../Domain/Entities/Walk"
import "../../../Domain/Entities/Location"
import "../../../Data/Repositories/WalkRepository"
import "../../../Data/Repositories/LocationRepository"

/// TrackWalkUseCase: Responsible for tracking real-time location during walks
/// Requirement: Service Execution (1.3 Scope/Core Features/Service Execution)
/// Supports live GPS tracking, route recording, and location-based features during dog walking sessions
class TrackWalkUseCase {
    
    // MARK: - Properties
    
    private let walkRepository: WalkRepository
    private let locationRepository: LocationRepository
    
    // MARK: - Initialization
    
    /// Initializes the TrackWalkUseCase with required repositories
    /// - Parameters:
    ///   - walkRepository: Repository for managing walk data
    ///   - locationRepository: Repository for managing location data
    init(walkRepository: WalkRepository, locationRepository: LocationRepository) {
        self.walkRepository = walkRepository
        self.locationRepository = locationRepository
    }
    
    // MARK: - Public Methods
    
    /// Tracks the real-time location of a walk and updates the route
    /// - Parameters:
    ///   - walkId: The ID of the walk being tracked
    ///   - latitude: The current latitude coordinate
    ///   - longitude: The current longitude coordinate
    func trackLocation(walkId: String, latitude: Double, longitude: Double) {
        Logger.info("Tracking location for walk \(walkId): lat=\(latitude), lon=\(longitude)")
        
        // Create location instance
        let location = Location(latitude: latitude, longitude: longitude)
        
        // Save location to repository
        Task {
            do {
                // Save location data
                let locationSaved = await locationRepository.updateLocation(location)
                guard locationSaved else {
                    Logger.error("Failed to save location for walk \(walkId)")
                    return
                }
                
                // Fetch current walk
                if let walk = try await walkRepository.fetchWalks().first(where: { $0.id == walkId }) {
                    // Add location to walk route
                    walk.addLocation(Walk.Location(latitude: latitude, longitude: longitude))
                    
                    // Save updated walk
                    try await walkRepository.updateWalk(walk)
                    Logger.info("Successfully updated walk \(walkId) with new location")
                } else {
                    Logger.error("Failed to fetch walk with ID \(walkId)")
                }
            } catch {
                Logger.error("Error tracking location: \(error.localizedDescription)")
            }
        }
    }
}

// MARK: - Error Handling Extension

extension TrackWalkUseCase {
    /// Custom error types for location tracking operations
    enum TrackingError: LocalizedError {
        case invalidWalkId
        case invalidCoordinates
        case locationSaveFailed
        case walkUpdateFailed
        
        var errorDescription: String? {
            switch self {
            case .invalidWalkId:
                return "Invalid walk ID provided"
            case .invalidCoordinates:
                return "Invalid location coordinates provided"
            case .locationSaveFailed:
                return "Failed to save location data"
            case .walkUpdateFailed:
                return "Failed to update walk with new location"
            }
        }
    }
}